### PR TITLE
Remove missing_tools config

### LINKED ISSUE #79249: Remove `missing_tools` config
The scope of this issue changed; see [here](https://github.com/rust-lang/rust/issues/79249#issuecomment-1415070926) for the current status.

Original issue description for historians:

---

config.toml lets you request tools be built, and it lets you ignore build failures with `dist.missing-tools = true`, but it doesn't let you specify that some packages *must* be built, while others are ok to miss.

Ideally, it would support either an optional-tools array, or missing-tools could take an array of tools that are ok to miss, so that this nuance could be expressed.

This is a little niche, but I'm running into this while hooking up a "build rust from source" option in firefox's CI, where rustfmt is optional but nice-to-have for lints. Currently I think the best option is to using `missing-tools = true` and check the output for warnings about skipped tools in the dist step. This is a bit fragile though, so it would be nice if this was built in.

@Mark-Simulacrum mentionned this would also be desirable for `miri` on stable/beta?

### PR DESCRIPTION
**Pull request by tharunsuresh-code**
Fixes https://github.com/rust-lang/rust/issues/79249

Request to check if run_cargo function has been appropriately modified and what changes need to be made

### CODE REVIEW

**Comment by ozkanonur** [comment]
Seems like changes made in this PR can't be passed in the pipelines currently. Please make sure all the pipelines gets passed successfully and then type `ready @rustbot` so we can review the changes.

File: src/bootstrap/tool.rs
Comment (tharunsuresh-code) [inline src/bootstrap/tool.rs:98]: Do we need both directories here?

File: src/bootstrap/tool.rs
Comment (ozkanonur) [inline src/bootstrap/tool.rs:90]: We shouldn't need `cargo_stamp` computition and `compile::run_cargo` execution. Did you try using `compile::stream_cargo` as before(but without the old logic inside of it's closure param)?

Comment (tharunsuresh-code) [reply src/bootstrap/tool.rs:90]: Do you mean something like this? 
`compile::stream_cargo(builder, cargo, vec![], &mut |_msg| return);`

This works! Though can you drop a small explanation when to choose which one? It is for my understanding.

File: src/bootstrap/tool.rs
Comment (ozkanonur) [inline src/bootstrap/tool.rs:52]: `is_optional_tool` should be completely removed from `ToolBuild`

File: src/bootstrap/tool.rs
Comment (ozkanonur) [inline src/bootstrap/tool.rs:86]:
```suggestion
        // Streamlined building all tools mandatorily
```

**Review by onur-ozkan — COMMENTED** [review]
Could you please also squash your commits into single one?

File: src/bootstrap/tool.rs
Comment (ozkanonur) [reply src/bootstrap/tool.rs:90]: > Do you mean something like this? `compile::stream_cargo(builder, cargo, vec![], &mut |_msg| return);`

Yes, I meant exactly that, but I see this was the particular reqeuest in the PR issue. I didn't see that earlier, sorry. This should stay :) 


> This works! Though can you drop a small explanation when to choose which one (between run and stream)? It is for my understanding.

You can use a callback closure in `stream_cargo` for processing the messages coming from cargo command's stdout. `run_cargo` also uses `stream_cargo` in it.

File: src/bootstrap/tool.rs
Comment (tharunsuresh-code) [reply src/bootstrap/tool.rs:52]: got it, done

File: src/bootstrap/tool.rs
Comment (tharunsuresh-code) [reply src/bootstrap/tool.rs:89]: alright, got it! Thanks! I have retained the code

**Comment by tharunsuresh-code** [comment]
> Could you please also squash your commits into single one?

Done!

File: src/bootstrap/tool.rs
Comment (ozkanonur) [inline src/bootstrap/tool.rs:89]: A small optimization:
```suggestion
        let out_dir = builder.cargo_out(compiler, self.mode, target);
        let cargo_out = out_dir.join(exe(tool, target));
        let bootstrap_tool_stamp = out_dir.join(".bootstrap-tool.stamp");
        let bin = builder.tools_dir(compiler).join(exe(tool, target));

        // Streamlined building all tools mandatorily
        compile::run_cargo(builder, cargo, vec![], &bootstrap_tool_stamp, vec![], false, false);
```

**Review by onur-ozkan — CHANGES_REQUESTED** [review]
This looks great! Just one more note before taking this into master.

**Comment by tharunsuresh-code** [comment]
> This looks great! Just one more note before taking this into master.

Sure, the code looks better!

**Review by onur-ozkan — APPROVED** [review]
thanks for the review fixes!

File: src/bootstrap/tool.rs
Comment (jyn514) [inline src/bootstrap/tool.rs:491]: This comment is now outdated.

File: src/bootstrap/tool.rs
Comment (jyn514) [inline src/bootstrap/tool.rs:171]: This check was removed and not replaced with anything. That seems ok to me, we don't check this for the compiler itself, but wanted to note that explicitly so it's easy to find in the future.

**Comment by matthiaskrgr** [comment]
Something broke miri in the rollup ^ and I can't see any changes to miri itself, wondering if it was this one...
@bors rollup=iffy

**Comment by tharunsuresh-code** [comment]
Any directions on how to check the error mentioned above?

**Comment by ozkanonur** [comment]
> Any directions on how to check the error mentioned above?

I sımply run `x buıld --stage 0` and then linked `stage0-sysroot` with rustup. And then executed the commands you can see in the `sh` outputs. I guess only difference between your and my case is I have `x86_64` machine.

**Comment by ozkanonur** [comment]
> Any directions on how to check the error mentioned above?

You can run `x dist` for targets which had errors in the pipelines. For windows specific one, don't forget to set `download-ci-llvm = false` because it failed in the llvm build.

The problem here might be that there is no fallback on building tools because we force them to be successfuly built.

cc @jyn514

**Comment by tharunsuresh-code** [comment]
So do I need to check this in a windows environment? Because my Linux compilation in local works just fine

**Comment by ozkanonur** [comment]
> So do I need to check this in a windows environment?

You can catch [this error](https://github.com/rust-lang-ci/rust/actions/runs/4158194624/jobs/7193240987) which is raised on linux environment.

> Because my Linux compilation in local works just fine.

I guess you run the `x86_64` compilations because I can repreduce the error.

Here is a little help for repreducing the error:

config.toml that is used in failed pipeline
```toml
changelog-seen = 2

[llvm]
static-libstdcpp = true
download-ci-llvm = false

[build]
extended = true
docs = false
print-step-timings = true
metrics = true
submodules = false
locked-deps = true
cargo-native-static = true
configure-args = ['--enable-extended', '--disable-docs']

[rust]
codegen-units-std = 1
remap-debuginfo = true
verbose-tests = true
dist-src = false
channel = "nightly"
debuginfo-level-std = 1

[dist]
compression-formats = ["xz"]
```

command that is used in failed pipeline
```sh
python3 x.py dist --host mips64el-unknown-linux-gnuabi64 --target mips64el-unknown-linux-gnuabi64
```

I tried this in ubuntu 22.04 container(because I can't get some of the required packages on my host system for this particular test) and I can confirm the exact same error raises.

**Comment by tharunsuresh-code** [comment]
Thanks for the detailed steps! I will try it out

**Comment by tharunsuresh-code** [comment]
From what I gather, it looks an issue with libffi-sys. It does not have any "mips" architecture defined in their `arch.rs` file here -> https://github.com/tov/libffi-rs/blob/master/libffi-sys-rs/src/arch.rs

Should I take it up with the maintainers of that repository asking to include mips and mips64 architecture? 

Specific error:

```shell
error[E0425]: cannot find value `FFI_TRAMPOLINE_SIZE` in this scope
   --> libffi-sys-rs/src/lib.rs:165:25
    |
165 |     pub tramp: [c_char; FFI_TRAMPOLINE_SIZE],
    |                         ^^^^^^^^^^^^^^^^^^^ not found in this scope
    |
note: these constants exist but are inaccessible
   --> libffi-sys-rs/src/arch.rs:34:9
```

**Comment by jyn514** [comment]
@tharunsuresh-code where do you see that error? when building which tool? I don't see it in the github actions logs.

**Comment by jyn514** [comment]
The LLVM error looks spurious so I think we can retry this PR: https://github.com/rust-lang/rust/issues/108227

@bors retry

**Comment by jyn514** [comment]
oh no I'm wrong, I missed the miri error @ozkanonur pointed out. Sorry for the noise.

@bors r-

**Comment by jyn514** [comment]
> Should I take it up with the maintainers of that repository asking to include mips and mips64 architecture?

Yes, that seems like the right thing to do :) mips64-unknown-linux-gnuabi64 is "Tier 2 with Host Tools" so theoretically it should have all these tools available.

cc @RalfJung @Mark-Simulacrum in case you have opinions on whether we should require Miri to be available on tier 2 targets; right now it's allowed to be missing.

**Comment by tharunsuresh-code** [comment]
> The LLVM error looks spurious so I think we can retry this PR: #108227
> 
> @bors retry

Thanks for pointing this out! I was trying to recreate the windows error and was having successful builds on my local windows machine :sweat_smile:

**Comment by RalfJung** [comment]
> in case you have opinions on whether we should require Miri to be available on tier 2 targets; right now it's allowed to be missing.

No strong opinion. Assuming that it doesn't put much of a burden on Miri itself (which I doubt it does; adding support for more *targets* is tricky but I don't quite see how more *host* triples could be a problem), I'd leave this to the people that work on these targets. Though it seems you did run into libffi issues... we could totally disable FFI support on hosts not supported by libffi.

However using Miri on those targets might require passing something like  `--target x86_64-unknown-linux-gnu` (which works on any host) to ensure maximal Miri support. That said, the Miri test suite actually checks `--target mips64-unknown-linux-gnuabi64` to ensure things work on a big-endian target, so that particular triple should work fine.

**Comment by jyn514** [comment]
> we could totally disable FFI support on hosts not supported by libffi

Hmm, how could we do that without hard coding the target name? I worry that we'll disable FFI for the target, it will get supported upstream, and then it'll stay unsupported in miri indefinitely because we don't notice.

**Comment by RalfJung** [comment]
Yeah we'd rely on someone reporting an issue to unlock this in Miri.

File: src/bootstrap/tool.rs
Comment (ehuss) [inline src/bootstrap/tool.rs:699]: This is intentionally set to be optional. These credential programs had some risk that they wouldn't compile in all environments. I suppose if you can get CI to pass on all targets, then it should be ok. But I want to point out that this is a risky change, and I'm a little nervous about doing this.

File: src/bootstrap/config.rs
Comment (ehuss) [inline src/bootstrap/config.rs:205]: Removing this will break anyone who has it set in their config. Is there a procedure for removing settings in such a way to alert people ahead-of-time that their config will no longer work?

File: src/bootstrap/tool.rs
Comment (ehuss) [reply src/bootstrap/tool.rs:215]: Why is this being removed? The duplicate artifact check is important for ensuring that some dependencies are not unnecessarily built multiple times which can increase CI time. There's a large amount of infrastructure (particularly the workspace-hack) that this would unravel.

Also, these changes seem unrelated to removing `missing-tools`. Is there a full description of what this PR is doing and why?

**Comment by jyn514** [comment]
https://github.com/tov/libffi-rs/issues/67 was fixed, might be worth rechecking if miri builds on mips now

**Comment by RalfJung** [comment]
MIPS is a tier 3 target now (https://github.com/rust-lang/rust/pull/115238), so if this PR was blocked on "Miri should built on all tier 2 targets" then MIPS is not even a concern any more.

**Pull request closed without merging**
